ARCHITECTURE.md

```diff
@@ -0,0 +1,241 @@
+the call.
+
+#### Login
+
+Logging in is an operation initiated by the Client. It may result in one of the following:
+- a redirection of the Resource Owner to the Issuer's authorization endpoint
+- a request by the Client to the Issuer's token endpoint
+Handlers for the login operation are located in `packages/*/src/login/oidc/oidcHandlers/*Handler.ts`.
+
+#### Incoming redirect
+
+The incoming redirect is an operation initiated by the Issuer.
+At the Issuer webpage, the Resource Owner authenticates (e.g. by entering a username
+and a password), after which the Issuer sends them to a webpage under the Client
+app's control (its `redirect_uri`), to which it appends some query parameters that
+it can use to complete the login flow. This is done when the developer calls 
+`handleIncomingRedirect`, and the Handlers for the incoming redirect are located
+in `packages/*/src/login/oidc/redirectHandler/*Handler.ts`.
+
+### Dependency injection
+
+An important architectural component of this library is dependency injection,
+implemented here using [TSyringe](https://github.com/Microsoft/tsyringe). Dependencies
+are declared in `packages/*/src/dependencies.ts`.
+
+#### Declaring order
+
+Something important to realize is that the order in which the dependencies are
+declared (for a given container) matters. Let's have a look at some code to make
+things clearer.
+
+```
+container.register<IOidcHandler>("browser:oidcHandler", {
+  useClass: AggregateOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: RefreshTokenOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: AuthorizationCodeWithPkceOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: ClientCredentialsOidcHandler,
+});
+```
+
+Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler
+Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler`
+and `ClientCredentialsOidcHandler` are its underlying handlers. When receiving a
+request, `AggregateOidcHandler` will first call to its instance of `RefreshTokenOidcHandler`
+to check if it can handle it. If so, the instance of `RefreshTokenOidcHandler`
+will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler`
+or `ClientCredentialsOidcHandler` will not be called. This means that it is
+important to declare the dependencies from the most specialized to the most generic,
+because if a fallback handler that can handle all requests is declared first,
+the other more specialized handlers will not be called.
+
+Note that the label for the containers of the aggregator and the underlying handlers
+differ (respectively `browser:oidcHandler` and `browser:oidcHandlers`, with an 's',
+in the above example). In this case, the order in which the dependencies are
+declared is irrelevant, because they do not relate to the same container. The
```

Comment:
For my own education -  What does "they do not relate to the same container" mean?  I see the same container registering the handlers.

Reply:
It's a concept really specific to TSyringe, so you don't need to worry too much about it :). What I'm trying to mean is that the order in which the `container` object registers dependencies isn't relevant  in the case they are registered with different labels. Maybe that's actually the way this should be phrased :D